### ISSUE TITLE
Restart waits for all processes to stop before starting any

### ISSUE THREAD

**Issue by mal** [Supervisor/supervisor#723]
It's not clear if this is by design, I can see it might be useful in some circumstances, however it would also be useful to restart each process as soon as it stops, perhaps behind a flag?

Here is an example. Processes `a`, `b` and `c` all stop immediately, while `d` stays in the `STOPPING` state for ~90 seconds. Below you can see the current and desired behavior:

Current Behavior | Desired Behavior
------------------|-------------------
`supervisor> restart all` | `supervisor> restart all`
`c: stopped` | `c: stopped`
`b: stopped` | `b: stopped`
`a: stopped` | `a: stopped`
[90 second wait] | `b: started`
`d: stopped` | `a: started`
`b: started` | `c: started`
`d: started` | [90 second wait]
`a: started` | `d: stopped`
`c: started` | `d: started`

**Comment by jvanasco** [comment]
I think it would make sense to have a flag if each item should be restarted in order:

| current | suggested |
|----|----|
| stop a | stop a |
| stop b | start a |
| stop c | stop b |
| stop d | start b |
| start a | stop c |
| start b | start c |
| start c | stop d |
| start d | start d |

The situation I've run into:  Our processes stop fast, they start slowly due to pre-caching and bottlenecks.  Doing a restart turns EVERYTHING off, then brings things back online one-by-one. This can create extended downtimes on the latter processes.

**Comment by Seldaek** [comment]
Related to this, it would be nice if `stop all` would actually stop all processes in parallel. Right now it seems they are stopped one by one and it blocks every time before signaling the next. As we have some jobs that take a while to stop gracefully it can result in restart taking a few minutes. If things were done in parallel (i.e. first signal all processes and mark them as stopping so they don't get restarted), then wait until all are stopped to consider the stop job done, that would help a lot. As a workaround I signal them all by hand first then call `restart all` but as the jobs are not marked stopped they can restart in the meantime so it's hardly a perfect solution.

**Comment by bharathwaaj** [comment]
There should be an option to restart one by one. I currently achieve like this.

`sudo supervisorctl restart gunicorn && sudo supervisorctl restart celeryd`

**Comment by kevinkreiser** [comment]
I've found a likely related issue with shutting down supervisord. Consider the following application `work.py` which listens for SIGTERM and upon hearing it, cleans up for 5 seconds before gracefully shutting down:

```python
#!/usr/bin/env python
import signal
import time
import threading
import sys

def cleanup(sig, frame):
  print('cleaning')
  time.sleep(5)
  global shutdown
  shutdown = True

def work(name):
  global shutdown
  while not shutdown:
    print('%s is working' % name)
    time.sleep(1)

if __name__ == '__main__':
  global shutdown
  shutdown = False
  signal.signal(signal.SIGTERM, cleanup)
  x = threading.Thread(target=work, args=(sys.argv[1],))
  x.start()
  x.join()
```

Then consider the following `supervisord.conf` in which we run multiple separate instances of the above application with different arguments (seems contrived but in real life you can imagine a bunch of different executables):

```ini
[supervisord]
logfile = /tmp/supervisord.log
loglevel = info

[inet_http_server]
port = 127.0.0.1:9001

[supervisorctl]
serverurl = http://127.0.0.1:9001

[rpcinterface:supervisor]
supervisor.rpcinterface_factory = supervisor.rpcinterface:make_main_rpcinterface

[program:foo]
command = ./work.py foo
process_name = foo

[program:bar]
command = ./work.py bar
process_name = bar
```

When you run this with `supervisord -n -c supervisord.conf` and you send SIGTERM to supervisord, you get the following output:

```
2021-03-01 19:16:27,696 INFO RPC interface 'supervisor' initialized
2021-03-01 19:16:27,696 CRIT Server 'inet_http_server' running without any HTTP authentication checking
2021-03-01 19:16:27,696 INFO supervisord started with pid 344951
2021-03-01 19:16:28,698 INFO spawned: 'bar' with pid 344953
2021-03-01 19:16:28,699 INFO spawned: 'foo' with pid 344954
2021-03-01 19:16:29,701 INFO success: bar entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2021-03-01 19:16:29,701 INFO success: foo entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2021-03-01 19:16:34,709 WARN received SIGTERM indicating exit request
2021-03-01 19:16:34,709 INFO waiting for bar, foo to die
2021-03-01 19:16:37,712 INFO waiting for bar, foo to die
2021-03-01 19:16:39,736 INFO stopped: foo (exit status 0)
2021-03-01 19:16:40,738 INFO waiting for bar to die
2021-03-01 19:16:43,741 INFO waiting for bar to die
2021-03-01 19:16:45,737 INFO stopped: bar (exit status 0)
```

The problem seems to be that supervisord sends SIGTERM to each program sequentially starting at the last one defined in the config and working its way up to the first one but it waits for the current one to exit before sending SIGTERM to the next one. We end up having to wait 10 seconds for everything to properly shutdown when we should have only had to wait for 5 or 6 seconds.

Basically its doing something like this:

```python
for program in programs:
  program.terminate() # send signal
  wait_for_terminated(program, timeout) # wait for death
  # make sure its dead
  if program.alive():
    program.kill()
```

When users expect it to be doing something like this:

```python
# send all signals
for program in programs:
  program.terminate()
# wait for all deaths
for program in programs:
  wait_for_terminated(program, timeout)
  # make sure its dead
  if program.alive():
    program.kill()
```

Do the maintainers consider it a problem that multiple programs are not being signaled to stop at the same time? Or perhaps there is an existing config option in supervisord that I am just not aware of? I looked at the code to see if I could make a PR but haven't figured out exactly where supervisord is sending the signals to the child processes yet. Oh I should also mention I'm on version 4.1.0 and it still exhibits this behavior.

**Comment by kevinkreiser** [comment]
I found a work around for at least my issue of not signalling shutdown at the same time, perhaps it also addresses the the OPs issue about starting/restarting as well I havent verified it. What you can do is you can use grouping in your supervisord config (ie. `[group:x]`). I have found that if you do this signals will indeed hit all processes at the same time.

Specifically, if you take my example above and add to the end of your config:

```ini
[group:baz]
programs=foo,bar
```

The two programs will get SIGTERM simultaneously and quit simultaneously as expected.

**Comment by akuzia** [comment]
I think I found out cause of the problem. In my case there was a slow `supervisorctl` reload with large ammount of tasks.

I've made a [demo repository with simple benchmark](https://github.com/akuzia/supervisord-slow-reload-fix).

TLDR from repo:

supervisorctls controller plugins [do_signal](https://github.com/Supervisor/supervisor/blob/8b83afdac70f140787fdb160d930f824a1f90a33/supervisor/supervisorctl.py#L918) separate process names and group names and signall them differenly.

RPC on the other side on receiving signal proceeds incoming signal to group processes via `make_all_func` [SignalProcessGroup](https://github.com/Supervisor/supervisor/blob/8b83afdac70f140787fdb160d930f824a1f90a33/supervisor/rpcinterface.py#L520), while processes get signalled one at a time.

Not sure, but this looks like it was made by design.
